packages/labs/virtualizer/src/test/scenarios/fixed-position-ancestor.test.ts

```diff
@@ -0,0 +1,119 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {array, ignoreBenignErrors} from '../helpers.js';
+import {LitVirtualizer} from '../../lit-virtualizer.js';
+import {grid} from '../../layouts/grid.js';
+import {styleMap} from 'lit/directives/style-map.js';
+import {expect, html, fixture} from '@open-wc/testing';
+
+interface NamedItem {
+  name: string;
+}
+
+const _100Items = array(100).map((_, i) => ({name: `Item ${i}`}));
+
+describe('Virtualizer behaves properly when it has a position: fixed ancestor', () => {
+  ignoreBenignErrors(beforeEach, afterEach);
+
+  // Regression test for https://github.com/lit/lit/issues/3904
+  it('should calculate its viewport correctly', async () => {
+    // Our outer container is overflow: hidden, so would normally be
+    // a "clipping ancestor" of our virtualizer. Since it has a
+    // width of zero, it would cause the virtualizer to be invisible.
+    const containerStyles = {
+      overflow: 'hidden',
+      width: '0',
+    };
+
+    // But since this scrolling element is position: fixed, it will
+    // be positioned relative to the document, essentially "popping out"
+    // of its zero-width parent. This means the virtualizer will be
+    // visible and have some space to work with, after all.
+    const scrollerStyles = {
+      position: 'fixed',
+      width: '200px',
+      height: '200px',
+    };
+
+    // We use the grid layout here because it conveniently will not
+    // render any items if it calculates that it doesn't have sufficient
+    // space, making our testing job easier.
+    const container = await fixture(html`
+      <div id="container" style=${styleMap(containerStyles)}>
+        <div id="scroller" style=${styleMap(scrollerStyles)}>
+          <lit-virtualizer
+            .layout=${grid({
+              itemSize: '100px',
+            })}
+            .items=${_100Items}
+            .renderItem=${({name}: NamedItem) => html`<div>${name}</div>`}
+          ></lit-virtualizer>
+        </div>
+      </div>
+    `);
+
+    const scroller = container.querySelector('#scroller')!;
+    expect(scroller).to.be.instanceOf(HTMLDivElement);
+    const virtualizer = container.querySelector('lit-virtualizer')!;
+    expect(virtualizer).to.be.instanceOf(LitVirtualizer);
+
+    // If virtualizer has properly ignored the zero-width ancestor of our
+    // fixed-position scroller, some children will be rendered; otherwise, not.
+    //
+    // In practice, we'll time out if we fail here because the `layoutComplete`
+    // promise will never be fulfilled.
+    await virtualizer.layoutComplete;
+    expect(virtualizer.textContent).to.contain('Item 0');
```

Comment:
Considering a failure's a failure so I'm okay either way, but what are your thoughts on using `pass` here since the failure message would be more explicit?

Reply:
I prefer to avoid using `pass` except in cases where we really can't avoid it; instead, I'd prefer to make sure `layoutComplete` is as reliable as possible and use that for cases where we want to check the state of things after a virtualizer renders.

That said, I was also bothered by the fact that this test failed with a timeout rather than something more akin to the failure you'd observe in manual testing, so I did a bit of digging. It turns out that there were two quirks in the `layoutComplete` implementation that resulted in the timeout in this case:

- When you provided a layout explicitly, rather than relying on the default layout to be dynamically loaded, the initial layout pass would complete before you had a chance to request the `layoutComplete` promise.

- If a layout pass resulted in no children being rendered (as in the failure case for this test), `layoutComplete` would never resolve.

I have made tweaks to address both of these quirks and pushed a new commit to this branch. There is a decent chance that making this change will require users to update some existing tests; specifically, if they are rendering a virtualizer and immediately checking the state of its initial render (without any delay or retry logic), they will need to insert an `await virtualizer.layoutComplete`. (I had to do this to two of our existing tests.) So we need to decide whether we want to go ahead (I'm inclined to) and, if so, whether we call this a breaking change for the Virtualizer release (I'm open to discussion).

Reply:
I'm torn. I think the change to `layoutComplete` is worth discussing but probably better to be split from this PR just so we can land the `position: fixed` scroller fix first.

Like the comment you added, it feels bad to introduce a delay purely for the sake of a promise whose main use is for testing.

Reply:
Yeah, that makes sense. I'll roll back that commit here and make the `layoutComplete` changes into a separate PR.